### DISCUSSION TITLE
[New Feature]  Add `padStart`, `padEnd`, `padZero` langlib functions to `string.bal`

### DISCUSSION THREAD

**Discussion by SandaruJayawardana** [ballerina-platform/ballerina-lang#35587] [General]
User can easily add padding in strings with this feature.
+ `padStart` - Adds padding to the start of a string

```
# Adds padding to the start of a string.
# Adds sufficient `padChar` characters at the start of `str` to make its length be `len`.
# If the length of `str` is >= `len`, returns `str`.
# 
# + str - the string to pad
# + len - the length of the string to be returned
# + padChar - the character to use for padding `str`; defaults to a space character
# + return - `str` padded with `padChar` 
public isolated function padStart(string str, int len, Char padChar = " ") returns string = external;
```
Ex:
> str - `"Ballerina"`
> len - `12`
> padChar - `"-"`
> return - `"---Ballerina"`
> Explanation: Length of the given string is 9. Target length is 12. Since 12 - 9 = 3 (> 0), we have to add 3 of `-` as the padding.

> str - `"Ballerina"`
> len - `9`
> padChar - `"-"`
> return - `"Ballerina"`
> Explanation: Length of the given string is 9. Target length is 9. Since 9 - 9 = 0 (<= 0), we don't need to add any padding.

+ `padEnd` - Adds padding to the end of a string
```
# Adds padding to the end of a string.
# Adds sufficient `padChar` characters to the end of `str` to make its length be `len`.
# If the length of `str` is >= `len`, returns `str`.
# 
# + str - the string to pad
# + len - the length of the string to be returned
# + padChar - the character to use for padding `str`; defaults to a space character
# + return - `str` padded with `padChar` 
public isolated function padEnd(string str, int len, Char padChar = " ") returns string = external;
```

Ex:
> str - `"Ballerina"`
> len - `12`
> padChar - `"-"`
> return - `"Ballerina---"`
> Explanation: Length of the given string is 9. Target length is 12. Since 12 - 9 = 3 (> 0), we have to add 3 of `-` as the padding.

> str - `"Ballerina"`
> len - `5`
> padChar - `"-"`
> return - `"Ballerina"`
> Explanation: Length of the given string is 9. Target length is 5. Since 5 - 9 = -4 (<= 0), we don't need to add any padding.

+ `padZero` - Adds padding to the start of a string
```
# Pads a string with zeros.
# The zeros are added at the start of the string, after a `+` or `-` sign if there is one.
# Sufficient zero characters are added to `str` to make its length be `len`.
# If the length of `str` is >= `len`, returns `str`.
# 
# + str - the string to pad
# + len - the length of the string to be returned
# + zeroChar - the character to use for the zero; defaults to ASCII zero `0`
# + return - `str` padded with zeros 
public isolated function padZero(string str, int len, Char zeroChar = "0") returns string = external;
```
Ex:
> str - `"100"`
> len - `5`
> zeroChar - 
> return - `"00100"`
> Explanation: Length of the given string is 3. Target length is 5. Since 5 - 3 = 2 (> 0), we have to add 2 of `0` as the padding.

> str - `"-100"`
> len - `5`
> padChar -
> return - `"-0100"`
> Explanation: Length of the given string is 4. Target length is 5. Since 5 - 4 = 1 (> 0), we have to add 1 of `0` as the padding.

Your input is much appreciated on this.

Lang issues: https://github.com/ballerina-platform/ballerina-lang/issues/35546, https://github.com/ballerina-platform/ballerina-lang/issues/35548

**Comment by SandaruJayawardana** [comment]
Since maximum length of Java `String` is `2^31 - 1`, we have to give an error when the given `len` is greater than `2^31 - 1`. It is an edge case. Is there any size limitation for ballerina string? otherwise, we have to give an error like `not yet supported`. right?

**Reply by gimantha** [reply]
> Since maximum length of Java `String` is `2^31 - 1`, we have to give an error when the given `len` is greater than `2^31 - 1`. It is an edge case. Is there any size limitation for ballerina string? otherwise, we have to give an error like `not yet supported`. right?

I think runtime error would work. This is a very edge case and implementation specific error. As per the spec I don't see any limitation for ballerina strings

**Reply by KavinduZoysa** [reply]
Please check https://github.com/ballerina-platform/ballerina-lang/pull/35598#discussion_r837579613

**Comment by KavinduZoysa** [comment]
@gimantha, @SandaruJayawardana, I saw that the implementation is done in java. Can't we implement it using ballerina?

**Reply by SandaruJayawardana** [reply]
Since we have used java string to keep ballerina string in jballerina, it won't work even we implement it using ballerina.
